lightning_examples/finetuning-scheduler/finetuning-scheduler.py

```diff
@@ -0,0 +1,754 @@
+#
+# ```python
+# trainer = Trainer(callbacks=[FinetuningScheduler(new_incarnation_mode=True)])
+# trainer.fit(..., ckpt_path="some/path/to/my_kth_best_checkpoint.ckpt")
+# ```
+#
+# To handle the edge case wherein one is resuming scheduled finetuning from a non-best checkpoint and the previous best checkpoints may not be accessible, setting ``FinetuningScheduler.new_incarnation_mode`` to
+# ``True`` as above will re-intialize the checkpoint state with a new best checkpoint at the resumption depth.
+
+# %% [markdown]
+# <div class="alert alert-warning">
+#
+# **Note:** Currently, _FinetuningScheduler_ only supports the following ``StrategyType``s:
+#
+# - ``DP``
+# - ``DDP``
+# - ``DDP_SPAWN``
+# - ``DDP_SHARDED``
+# - ``DDP_SHARDED_SPAWN``
+#
+# </div>
+
+# %% [markdown]
+# <div id="superglue"></div>
+#
+# ## Scheduled Finetuning For SuperGLUE
+#
+# The following example demonstrates the use of ``FinetuningScheduler`` to finetune a small foundational model on the [RTE](https://huggingface.co/datasets/viewer/?dataset=super_glue&config=rte) task of [SuperGLUE](https://super.gluebenchmark.com/). Iterative early-stopping will be applied according to a user-specified schedule.
+#
+
+# %%
+import os
+import warnings
+from datetime import datetime
+from importlib import import_module
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import datasets
+import pytorch_lightning as pl
+import torch
+from pytorch_lightning.loggers.tensorboard import TensorBoardLogger
+from pytorch_lightning.utilities import rank_zero_warn
+from pytorch_lightning.utilities.cli import CALLBACK_REGISTRY, _Registry
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from torch.utils.data import DataLoader
+from transformers import AutoConfig, AutoModelForSequenceClassification, AutoTokenizer
+from transformers import logging as transformers_logging
+
+# %%
+# a couple helper functions to prepare code to work with a user module registry
+MOCK_REGISTRY = _Registry()
+
+
+def mock_register_module(key: str, require_fqn: bool = False) -> List:
+    if key.lower() == "finetuningscheduler":
+        mod = import_module("pytorch_lightning.callbacks.finetuning_scheduler")
+        MOCK_REGISTRY.register_classes(mod, pl.callbacks.Callback)
+    else:
+        raise MisconfigurationException(f"user module key '{key}' not found")
+    registered_list = []
+    # make registered class available by unqualified class name by default
+    if not require_fqn:
+        for n, c in MOCK_REGISTRY.items():
+            globals()[f"{n}"] = c
+        registered_list = ", ".join([n for n in MOCK_REGISTRY.names])
+    else:
+        registered_list = ", ".join([c.__module__ + "." + c.__name__ for c in MOCK_REGISTRY.classes])
+    print(f"Imported and registered the following callbacks: {registered_list}")
+
+
+def instantiate_registered_class(init: Dict[str, Any], args: Optional[Union[Any, Tuple[Any, ...]]] = None) -> Any:
```

Comment:
Just personal feeling, but I think `instantiate_registered_class` make the tutorial more opaque to readers.

Reply:
Great suggestion. I think leaving this in the ``LightningCLI`` example in the module remains worthwhile but it was just adding an unnecessary layer of abstraction for this example. Thanks for all your valuable insight/feedback!